File: ts-src/file-history/cleanup.ts
import { readdir, stat, unlink } from "node:fs/promises";
import { join } from "node:path";
import { getFileHistoryDir } from "../config/paths.ts";

export interface FileHistoryCleanupConfig {
	maxSize: number; // bytes, default 100MB
	projectPath?: string;
}

interface CleanupStats {
	filesRemoved: number;
	bytesFreed: number;
}

const DEFAULT_MAX_SIZE = 100 * 1024 * 1024;

/**
 * Clean up old file history backups.
 * If total size exceeds maxSize, removes oldest versions first.
 */
export async function runFileHistoryCleanup(config?: Partial<FileHistoryCleanupConfig>): Promise<CleanupStats> {
	const maxSize = config?.maxSize ?? DEFAULT_MAX_SIZE;
	const baseDir = getFileHistoryDir(config?.projectPath);

	const stats: CleanupStats = { filesRemoved: 0, bytesFreed: 0 };

	let entries: string[];
	try {
		entries = await readdir(baseDir);
	} catch {
		return stats;
	}

	// Collect all backup files across all UUID dirs
	interface BackupInfo {
		path: string;
		version: number;
		size: number;
	}
	const allBackups: BackupInfo[] = [];

	for (const entry of entries) {
		if (entry === "meta.json") continue;
		const dirPath = join(baseDir, entry);
		let dirStat: Awaited<ReturnType<typeof stat>>;
		try {
			dirStat = await stat(dirPath);
		} catch {
			continue;
		}
		if (!dirStat.isDirectory()) continue;

		let files: string[];
		try {
			files = await readdir(dirPath);
		} catch {
			continue;
		}

		for (const f of files) {
			const match = f.match(/^v(\d+)\.bak$/);
			if (!match) continue;
			const version = Number.parseInt(match[1] as string, 10);
			const fullPath = join(dirPath, f);
			const info = await stat(fullPath);
			allBackups.push({ path: fullPath, version, size: info.size });
		}
	}

	// If total size exceeds maxSize, remove oldest versions first
	let totalSize = allBackups.reduce((sum, b) => sum + b.size, 0);
	if (totalSize > maxSize) {
		allBackups.sort((a, b) => a.version - b.version); // oldest first
		for (const backup of allBackups) {
			if (totalSize <= maxSize) break;
			await unlink(backup.path);
			stats.filesRemoved++;
			stats.bytesFreed += backup.size;
			totalSize -= backup.size;
		}
	}

	return stats;
}

File: src/file_history/cleanup.rs
//! Disk-budget cleanup: drop oldest backups when over `max_size`.

use std::path::PathBuf;

use regex::Regex;

use crate::config::paths::get_file_history_dir;

#[derive(Debug, Clone)]
pub struct CleanupConfig {
    /// bytes — defaults to 100 MiB
    pub max_size: u64,
    pub project_path: Option<String>,
}

impl Default for CleanupConfig {
    fn default() -> Self {
        Self {
            max_size: 100 * 1024 * 1024,
            project_path: None,
        }
    }
}

#[derive(Debug, Default, Clone)]
pub struct CleanupStats {
    pub files_removed: u64,
    pub bytes_freed: u64,
}

struct BackupInfo {
    path: PathBuf,
    version: u32,
    size: u64,
}

pub fn run_file_history_cleanup(config: CleanupConfig) -> CleanupStats {
    let mut stats = CleanupStats::default();
    let base_dir = get_file_history_dir(config.project_path.as_deref());
    let entries = match std::fs::read_dir(&base_dir) {
        Ok(e) => e,
        Err(_) => return stats,
    };
    let re = Regex::new(r"^v(\d+)\.bak$").unwrap();

    let mut all_backups: Vec<BackupInfo> = Vec::new();
    for entry in entries.flatten() {
        let name = entry.file_name();
        if name == "meta.json" {
            continue;
        }
        let dir_path = entry.path();
        if !dir_path.is_dir() {
            continue;
        }
        let inner = match std::fs::read_dir(&dir_path) {
            Ok(r) => r,
            Err(_) => continue,
        };
        for f in inner.flatten() {
            let fname = f.file_name().to_string_lossy().into_owned();
            let caps = match re.captures(&fname) {
                Some(c) => c,
                None => continue,
            };
            let version: u32 = match caps.get(1).and_then(|m| m.as_str().parse().ok()) {
                Some(v) => v,
                None => continue,
            };
            let size = f.metadata().map(|m| m.len()).unwrap_or(0);
            all_backups.push(BackupInfo {
                path: f.path(),
                version,
                size,
            });
        }
    }

    let mut total: u64 = all_backups.iter().map(|b| b.size).sum();
    if total > config.max_size {
        all_backups.sort_by_key(|b| b.version); // oldest first
        for backup in all_backups {
            if total <= config.max_size {
                break;
            }
            if std::fs::remove_file(&backup.path).is_ok() {
                stats.files_removed += 1;
                stats.bytes_freed += backup.size;
                total = total.saturating_sub(backup.size);
            }
        }
    }
    stats
}
